docs/HowToFAQ/FAQMapExtensionality.md

```diff
@@ -0,0 +1,114 @@
+---
+title: I have a lemma and later an assert stating the postcondition of the lemma, but it fails to prove. Why and how do I fix it?
+---
+
+## Question: I have a lemma and later an assert stating the postconditio nof the lemma, but it fails to prove. Why and how do I fix it?
+
+I have this lemma
+```dafny
+    lemma MapKeepsCount<X,Y,Z>(m : map<X,Y>, f : (X) -> Z)
+      requires forall a : X, b : X :: a != b ==> f(a) != f(b)
+      ensures |m| == |map k <- m.Keys | true :: f(k) := m[k]|
+```
+and later on this code
+```dafny
+MapKeepsCount(authSchema, k => Paths.SimpleCanon(tableName, k));
+assert |authSchema| == |map k <- authSchema.Keys | true :: Paths.SimpleCanon(tableName, k) := authSchema[k]|;
+```
+
+The final assert fails, even though it exactly matches the ensures of the lemma.
+What am I missing? 
+
+If the lemma is an axiom, one can try to assume the post condition right before the assertion. 
+But that failed in this case
+
+```dafny
+MapKeepsCount(authSchema, k => Paths.SimpleCanon(tableName, k));
+assume |authSchema| == |map k <- authSchema.Keys | true :: Paths.SimpleCanon(tableName, k) := authSchema[k]|;
+assert |authSchema| == |map k <- authSchema.Keys | true :: Paths.SimpleCanon(tableName, k) := authSchema[k]|;
+```
+
+Which makes no sense.
+I even put the function into a variable, and this still fails
+```dafny
+assume |authSchema| == |map k <- authSchema.Keys :: fn(k) := authSchema[k]|;
+assert |authSchema| == |map k <- authSchema.Keys :: fn(k) := authSchema[k]|;
+```
+
+## Answer:
+
+The explanation is a little involved, and in the end gets into a weakness of Dafny. But I also have a workaround. Get some popcorn. Here goes:
+
+To prove that the `|map …|` expression in the specification has the same value as the `|map …|` expression in the code, 
+the verifier would either have to compute the cardinality of each map (which it can’t do, because m.Keys is symbolic and could stand for any size set) 
```

Comment:
```suggestion
the verifier would either have to compute the cardinality of each map (which it can’t do, because `m.Keys` is symbolic and could stand for any size set) 
```